File: backend/src/test/scala/com/skewwhiffy/auraltester/scales/ScaleTest.scala
package com.skewwhiffy.auraltester.scales

import com.skewwhiffy.auraltester.notes.AbsoluteNote
import com.skewwhiffy.auraltester.testutils.TestData.noteFactories._
import org.scalatest.flatspec.AnyFlatSpec

class ScaleTest extends AnyFlatSpec {
  it should "can instantiate major scale" in {
    val expected = internalNotation.getNotes("C D E F G A B c")

    val actual = new Scale(AbsoluteNote.middleC, scaleTypeFactory.major, ScaleDirection.ascending)

    assert(actual.notes == expected)
  }

  it should "can instantiate minor harmonic scale" in {
    val expected = internalNotation.getNotes("D E F G A Bb c# d")

    val actual = new Scale(
      internalNotation.getNote("D"),
      scaleTypeFactory.minorHarmonic,
      ScaleDirection.ascending
    )

    assert(actual.notes == expected)
  }

  it should "can instantiate minor melodic ascending scale" in {
    val expected = internalNotation.getNotes("E F# G A B c# d# e")

    val actual = new Scale(
      internalNotation.getNote("E"),
      scaleTypeFactory.minorMelodic,
      ScaleDirection.ascending
    )

    assert(actual.notes == expected)
  }

  it should "can instantiate minor melodic descending scale" in {
    val expected = internalNotation.getNotes("a g f e d c B A")

    val actual = new Scale(
      internalNotation.getNote("A"),
      scaleTypeFactory.minorMelodic,
      ScaleDirection.descending
    )

    assert(actual.notes == expected)
  }
}

File: backend/src/test/kotlin/com/skewwhiffy/notation/model/scale/ScaleTest.kt
package com.skewwhiffy.notation.model.scale

import com.skewwhiffy.notation.model.note.AbsoluteNote
import com.skewwhiffy.test.util.TestData
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test

class ScaleTest {
  @Test
  fun `instantiate major scale`() {
    val expected = TestData.noteFactories.internalNotation.getNotes("C D E F G A B c")

    val actual = Scale(
      AbsoluteNote.middleC,
      TestData.noteFactories.scaleTypeFactory.major,
      ScaleDirection.ASCENDING
    )

    assertThat(actual.notes).isEqualTo(expected)
  }

  @Test
  fun `instantiate minor harmonic scale`() {
    val expected = TestData.noteFactories.internalNotation.getNotes("D E F G A Bb c# d")

    val actual = Scale(
      TestData.noteFactories.internalNotation.getNote("D"),
      TestData.noteFactories.scaleTypeFactory.minorHarmonic,
      ScaleDirection.ASCENDING
    )

    assertThat(actual.notes).isEqualTo(expected)
  }

  @Test
  fun `instantiate minor melodic ascending scale`() {
    val expected = TestData.noteFactories.internalNotation.getNotes("E F# G A B c# d# e")

    val actual = Scale(
      TestData.noteFactories.internalNotation.getNote("E"),
      TestData.noteFactories.scaleTypeFactory.minorMelodic,
      ScaleDirection.ASCENDING
    )

    assertThat(actual.notes).isEqualTo(expected)
  }

  @Test
  fun `instantiate minor melodic descending scale`() {
    val expected = TestData.noteFactories.internalNotation.getNotes("a g f e d c B A")

    val actual = Scale(
      TestData.noteFactories.internalNotation.getNote("A"),
      TestData.noteFactories.scaleTypeFactory.minorMelodic,
      ScaleDirection.DESCENDING
    )

    assertThat(actual.notes).isEqualTo(expected)
  }
}